Challonge

tordsk opened the pull request:
### Status

- [ ] Done
- [x] Work in progress
- [x] Needs testing from others

### Description
Made a prototype of tournament (challonge) bracket integration. 

Users join the competition, can check who their next opponent is, can register the score of the match.

Admins have a view of every playing game, can then see the registered score for each player (team).
The admin can select a winner and submit the score to challonge, the bracket will be updated and a new match will appear for the users.  

### Checklist

- [ ] The code is linted
- [ ] My changes requires change to the documentation.
- [ ] I have updated the documentation.
- [ ] I have introduced changes to build configuration.


### Model Changes
Model changes to competition:
- Participant: Added field: cid (Charfield), which is the challonge ID for a participant in the 
- New model: Match: Represents a match in challonge. Matches with state='open' are "playing" matches
-

### Breaking Changes

- added -e git+http://github.com/ZEDGR/pychal@db78a2572e258b1f174f5574a63ff2319df3e5cb#egg=pychal to requirements
 
-

### Other
The 'Start Competition' method/view, is really slow because it sends 4 requests to the challonge api
Submit score is also slow

### Screenshot (if appropriate)

larserikgk commented:
Very nice functionality. Some initial issues (some of which you are probably aware) and questions that I would like you to address;

- Credentials need to be moved to settings from the view
- Naming is based on time + activity title. This can be troublesome as challonge doesn't allow all characters in urls (CS:GO doesn't work. Maybe lan name + competition name is better?
- The way the reporting works now, an admin has to approve every result. Maybe the match should be auto approved when both teams report the same result? That way an admin only has to interact if there is disagreement between teams.
- The view 'start_compo' should have a try/catch for errors. If creation of the bracket fails, you are shown a stracktrace and 'Tournament has started!' message when you return to the view.
- This is a bit of work, and maybe beyond scope, but maybe the admin should have his own view of the competition? Or maybe an admin panel of some sort? It would be nice to choose 'tournament_type' for example.

tordsk commented:
== Update == 
- moved credentials to settings
- challonge_url is cleaned so it only contains alphanumeric characters
- automatic submission of matches to challonge implemented. Only when the score is decisive and both party agrees. Otherwise an admin has to submit the result. 
- added some try clause, but will probably need more testing
- not even going to start on this one, but good idea

== Other == 
- Can alter the competition model with a choice field to select single or double elimination, have only tested with single elim at this point.
- Should add a way of editing a match after submission. e.g. reopen a match

larserikgk commented on apps/competition/views.py:
```diff
@@ -331,43 +347,40 @@ def translate_competitions(competitions):
+            competition.status = 3
+            messages.success(request, 'Tournament has started!')
+            challonge.set_credentials(settings.CHALLONGE_API_USERNAME, settings.CHALLONGE_API_KEY)
+            url = str(int(time.time())) + competition.activity.title
```
I still think this might be better with lan.title + competition.activity.title here. It makes it easier to browse the different competitions on the CaG-challonge user and if we want to implement seeding later, it is easier to find previous tournaments programmatically. Is there a better reason to use time.time?

tordsk replied:
- Challonge_url
The challonge url has to be unique, and it is true that it would be unique for each lan unless someone creates a tournament on challonge with the url (e.g. StudLanvåren2012CSGO) before we start the tournament (in this case we need code to manually input challonge_url instead of only generating it), or if we have to restart the tournament for some reason (we would need to log into the challonge account, and delete the tourney before re-starting)

- Challonge title
We can change the title into whatever is useful, think this is what you mean ?
 `url = re.sub('[^0-9a-zA-Z]+', '', url)
            competition.challonge_url = url
            challonge.tournaments.create(competition.activity.title, url, tournament_type="single elimination")`

The competition.activity.title can be changed to lan.title + ':' + competition.activity.title which would make the challonge overview a bit more easy to read

larserikgk commented on apps/competition/views.py:
```diff
@@ -331,43 +347,40 @@ def translate_competitions(competitions):
+
+            update_match_list(competition)
+        except:
+            pass
```
This should probably return some kind of message to the user so we know something went wrong.

larserikgk commented on apps/competition/views.py:
```diff
@@ -331,43 +347,40 @@ def translate_competitions(competitions):
+            url = str(int(time.time())) + competition.activity.title
+            url = re.sub('[^0-9a-zA-Z]+', '', url)
+            competition.challonge_url = url
+            challonge.tournaments.create(competition.activity.title, url, tournament_type="single elimination")
```
Would like to see this as user input as we usually use double elim. As you pointed out, this can be put in the competition model.

larserikgk commented on templates/competition/competition.html:
```diff
@@ -38,10 +40,67 @@
+                                    {% csrf_token %}
+                                    <div class="form-actions center-on-collapse">
+                                         <select class="form-control pull-left" name="player1score" id="player1score">
+                                            <option value="1">1</option>
```
These values need to be set programmatically in the competition model as the values vary from game to game. LoL might have the score 1-0 and CS:GO might have the score 16-13.

tordsk replied:
Yes, initially my idea was just to make participants register the final score of the match (e.g. 1-0) in a bo1, (2-0) in a bo3 , (5-4) in a bo5 etc. Do we want (need?) to make the participants register score for each round ?

tordsk replied:
Yes the values should be done properly. On scoring though: Initially i thought that we should make the participants register the final score of the match, e.g. (1-0) in a bo1, (2-0) in bo3, and (4-5) in bo5. Do we want (need?) to register score for each round? In the case for CS:GO it would still be (1-0) in a bo1 right ?

larserikgk commented on requirements/requirements.txt:
```diff
@@ -0,0 +1,10 @@
+Django==1.7.1
+django-markdown-deux==1.0.4
+django-translatable==0.1.1
+reportlab
```
This is not needed.

larserikgk commented on requirements.txt:
```diff
@@ -0,0 +1,14 @@
+markdown2==2.3.2
+olefile==0.43
+Pillow==4.0.0
+-e git+http://github.com/ZEDGR/pychal@db78a2572e258b1f174f5574a63ff2319df3e5cb#egg=pychal
```
The new requirement should be added to requirements/base.txt instead, I think.

larserikgk commented:
This feature is probably not fine polished yet, but I think it is safe and stable enough to test more in production. Thoughts @MathiasFl @tordsk ?

tordsk commented on apps/competition/views.py:
```diff
@@ -365,6 +359,14 @@ def start_compo(request, competition_id):
+            url = unicode(competition.lan) + unicode(competition.activity)
+            url = re.sub('[^0-9a-zA-Z]+', '', url)
+            challonge.set_credentials(settings.CHALLONGE_API_USERNAME, settings.CHALLONGE_API_KEY)
+            challonge.tournaments.create(competition.activity.title, url, tournament_type="single elimination")
```
`url = unicode(competition.lan) + unicode(competition.activity)`
Still think there should be some kind of randomness to the url. 
Start compo can now be trolled by someone creating a tournament on challonge with the lan name + game name which isn't too far fetched. There is no easy recovery from this scenario, at least with the code as it is now.

larserikgk replied:
This is a good point. I will add either the date like it was or some other solution.
